### DISCUSSION TITLE
Performance stats

### DISCUSSION THREAD

**Discussion by Toflar** [loupe-php/loupe#17] [Show and tell]
People have asked me for performance stats which I cannot provide at the moment because I simply don't have any data yet :-)

If you're up to it, feel free to take on this task and run some tests. As a dataset, why not re-using what MeiliSearch provides? https://www.meilisearch.com/movies.json

You'll notice that indexing will take a loooong time but searching should hopefully be pretty quick - of course not MeiliSearch level but quick :)

**Comment by fiedsch** [comment]
Just tested some scenarios using the `movies.json` data and taking the `time` it took to run the program:

# Step one: Populate the database

create loupe database and populate with data

## (1) ->withSearchableAttributes() default, i.e. everything is indexed
```
real	17m4.271s
user	16m41.602s
sys	0m21.084s
database size is 202M
```

## (2) ->withSearchableAttributes(['title', 'overview'])
```
real	10m58.211s
user	10m44.493s
sys	0m12.995s
database size is 146M
```

## (3) like (1) but only with a subset of the data (first 1000 records of `movies.json`)
```
real	0m30.082s
user	0m29.893s
sys	0m0.176s
database size is 8M
```

## (4) like (2) but only with a subset of the data (first 1000 records of `movies.json`)
```
real	0m19.820s
user	0m19.659s
sys	0m0.141s
database size is 6M
```

# Step two: Query the already created loupe database


## database from (1)
```
real	0m51.997s
user	0m45.304s
sys	0m6.652s
```

## database from (2)
```
real	0m42.889s
user	0m37.169s
sys	0m5.685s
```


## database from (3)
```
real	0m0.194s
user	0m0.166s
sys	0m0.024s
```

## database from (4)
```
real	0m0.154s
user	0m0.130s
sys	0m0.020s
```


## "Findings" 

* the size of data should not be too large (I did not test with different settings of [Alphabet size and index length](https://github.com/loupe-php/loupe/blob/main/docs/configuration.md#alphabet-size-and-index-length) though.
* with large data sets it helps to reduce what is being indexed by explicitly setting columns using `withSearchableAttributes()`
* once the data size is moderate, speed is high and it does not matter too much if you index all or just some attributes. This might of course be different for data wit a lot of attributes.

**Comment by Toflar** [comment]
I did some more tests and research.

## Indexing

Of course, this very much depends on the amount of data. If you index the entire 32k movies from the `movies.json` it takes about **30s per 1000 documents**. I'm pretty sure the community can improve this by splitting indexing tasks or making good use of SQLite features I didn't know about. However, we should also consider that SQLite and PHP are single threaded, so stuff like parallelism that is used for other search engines is one thing we should probably not expect from Loupe.

And you should always configure the `searchableAttributes`. There's no need to waste time on indexing attributes you're never going to search on. 

## Searching

Again, this also very much depends on the way you're searching and whether typo tolerance is enabled or not.

First of all: I've found a few performance bugs in both, Loupe as well as my State Set Index library. So make sure you checkout the `0.2` branch (haven't tagged it yet) and also update the dependencies 😊 

### Benchmark 1

Searching 32k movies for `nemo` (which - because it's just 4 letters - will have typo tolerance disabled by default) ordered by relevance:

```php
$searchParameters = SearchParameters::create()
    ->withQuery('nemo')
;
```

**Processing time: 85 ms**

### Benchmark 2

Searching 32k movies for `nemo` (which - because it's just 4 letters - will have typo tolerance disabled by default) ordered by relevance but only searching the `title` attribute (new feature in 0.2):

```php
$searchParameters = SearchParameters::create()
    ->withQuery('nemo')
    ->withAttributesToSearchOn(['title'])
;
```

**Processing time: 36 ms**

### Benchmark 3

Okay, let's see what typo tolerance does. It must slow down the query considerably. Let's say we have a typo in `mysterious` and we write `mysterios`:

```php
$searchParameters = SearchParameters::create()
    ->withQuery('mysterios')
;
```

**Processing time: 1820 ms**

I consider this to be **amazingly fast**! It's a super broad search on 32k movies with just PHP and SQLite 🤯 
Also, most of the 1.8s is spent to calculate the possible states of the state set index algorithm, so I'm fairly sure there is potential to improve.

### Benchmark 4

Let's do the same as benchmark 3 but only have it search the `title` attribute again:

```php
$searchParameters = SearchParameters::create()
    ->withQuery('mysterios')
    ->withAttributesToSearchOn(['title'])
;
```

**Processing time: 1476 ms**

Even better! Remember: It's still searching 32k movies with typo tolerance enabled! This is probably not even a real world scenario. You'd likely filter for other attributes and narrow down the 32k movies in the first place.

## Conclusion

When it comes to indexing: Loupe is slow. You should calculate with about 35 documents per second. If you want to give it a try to make it faster, please do!

In terms of searching, I consider Loupe to be **amazingly fast**! I'm over the moon with the results for this early version. Of course, I did lots of performance testing before I published it so I ironed out a lot of mistakes but if you'd asked me at the beginning of the year whether writing a typo tolerant search engine with just PHP and SQLite that searches 32k documents in under 2 seconds was possible, I would've sure said "no way" 🥳 

And I'm fairly sure we can optimize all sorts of things and make it even faster out of the box 😊

**Comment by Toflar** [comment]
I did some more performance tests and optimizations and v0.2 will be able to search the 32k movies in under 400ms 🥳 I've added this to the README in https://github.com/loupe-php/loupe/commit/fc6f35b14a5b9d67c3591e2c2536d144a2aa86c9 and converted this issue to a GitHub discussion.
